File: problems-adapt/0115_life-one-generation/solution_state_bits.py
from typing import List, Optional


class Solution:
    def nextGeneration(self, board: List[List[int]]) -> List[List[int]]:
        m, n = len(board), len(board[0])
        dirs = ((-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1))
        for r in range(m):
            for c in range(n):
                live = 0
                for dr, dc in dirs:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < m and 0 <= nc < n and board[nr][nc] in (1, 2):
                        live += 1
                if board[r][c] == 1 and (live < 2 or live > 3):
                    board[r][c] = 2  # live -> dead
                elif board[r][c] == 0 and live == 3:
                    board[r][c] = 3  # dead -> live
        for r in range(m):
            for c in range(n):
                board[r][c] = 1 if board[r][c] in (1, 3) else 0
        return board

File: problems/0201-0300/0289_game-of-life/solution_state_bits.go
func gameOfLife(board [][]int) [][]int {
	m := len(board)
	n := len(board[0])
	dirs := [8][2]int{{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}}
	for r := 0; r < m; r++ {
		for c := 0; c < n; c++ {
			live := 0
			for _, d := range dirs {
				nr, nc := r+d[0], c+d[1]
				if nr >= 0 && nr < m && nc >= 0 && nc < n && (board[nr][nc] == 1 || board[nr][nc] == 2) {
					live++
				}
			}
			if board[r][c] == 1 && (live < 2 || live > 3) {
				board[r][c] = 2 // live -> dead
			} else if board[r][c] == 0 && live == 3 {
				board[r][c] = 3 // dead -> live
			}
		}
	}
	for r := 0; r < m; r++ {
		for c := 0; c < n; c++ {
			if board[r][c] == 1 || board[r][c] == 3 {
				board[r][c] = 1
			} else {
				board[r][c] = 0
			}
		}
	}
	return board
}
